components/fxa-client/src/lib.rs

```diff
@@ -2,597 +2,1268 @@
+//! something like:
+//!
+//! * Out-of-band, register your application with the Firefox Accounts service,
+//!   providing an OAuth `redirect_uri` controlled by your application and
+//!   obtaining an OAuth `client_id`.
+//!
+//! * On application startup, create a [`FirefoxAccount`] object to represent the
+//!   signed-in state of the application.
+//!     * On first startup, a new [`FirefoxAccount`] can be created by calling
+//!       [`FirefoxAccount::new`] and passing the application's `client_id`.
+//!     * For subsequent startups the object can be persisted using the
+//!       [`to_json`](FirefoxAccount::to_json) method and re-created by
+//!       calling [`FirefoxAccount::from_json`].
+//!
+//! * When the user wants to sign in to your application, direct them through
+//!   a web-based OAuth flow using [`begin_oauth_flow`](FirefoxAccount::begin_oauth_flow)
+//!   or [`begin_pairing_flow`](FirefoxAccount::begin_pairing_flow); when they return
+//!   to your registered `redirect_uri`, pass the resulting authorization state back to
+//!   [`complete_oauth_flow`](FirefoxAccount::complete_oauth_flow) to sign them in.
+//!
+//! * Display information about the signed-in user by using the data from
+//!   [`get_profile`](FirefoxAccount::get_profile).
+//!
+//! * Access account-related services on behalf of the user by obtaining OAuth
+//!   access tokens via [`get_access_token`](FirefoxAccount::get_access_token).
+//!
+//! * If the user opts to sign out of the application, calling [`disconnect`](FirefoxAccount::disconnect)
+//!   and then discarding any persisted account data.
+
 #![allow(unknown_lints)]
 #![warn(rust_2018_idioms)]
-
-use crate::{
-    commands::send_tab::SendTabPayload,
-    device::Device,
-    oauth::{AuthCircuitBreaker, OAuthFlow, OAUTH_WEBCHANNEL_REDIRECT},
-    scoped_keys::ScopedKey,
-    state_persistence::State,
-};
-pub use crate::{
-    config::Config,
-    error::*,
-    oauth::{AccessTokenInfo, IntrospectInfo, RefreshToken},
-    profile::Profile,
-    telemetry::FxaTelemetry,
-};
 use serde_derive::*;
-use std::{
-    cell::RefCell,
-    collections::{HashMap, HashSet},
-    sync::Arc,
-};
-use url::Url;
-
-#[cfg(feature = "integration_test")]
-pub mod auth;
-mod commands;
-mod config;
-pub mod device;
-pub mod error;
-pub mod ffi;
-mod http_client;
-pub mod migrator;
-// Include the `msg_types` module, which is generated from msg_types.proto.
-pub mod msg_types {
-    include!("mozilla.appservices.fxaclient.protobuf.rs");
+use std::collections::HashMap;
+use std::convert::{TryFrom, TryInto};
+use thiserror::Error;
+
+// All the implementation details live in this "internal" module.
+// Aspirationally, I'd like to make it private, so that the public API of the crate
+// is entirely the same as the API exposed to consumers via UniFFI. That's not
+// possible right now because some of our tests/example use features that we do
+// not currently expose to consumers. But we should figure out how to expose them!
```

Comment:
Just drawing attention to this aspirational comment for the future here. I think it would be good hygiene to have the UniFFIed interface be the *entire* public interface of the crate, so that e.g. our Rust examples use the same API surface as our eventual consumers. But that was a bigger refactor than I wanted to get into here, because some of our examples/tests use functionality that was not exposed over the FFI.

Reply:
That's a great aspiration. As mentioned above, the `URL/Guid <-> String` type things might require some uniffi changes, or maybe it already has what we need there - but yeah, not sure how much we should try and insist on that.